File: src/storystore.py
# storystore - HA python_script for persistant logging and storing.
# This service uses an entity (ENT) for storing records with time stamp and a set of strings.
# The records are separated by ';' and put as an attribute (by default named 'story').
# The entity's state is updated with the time stamp.
# The service stores a maximum of MAX records. Latest record is put last and the oldest one is removed when necessary.

# Split records on ;
# Split record on |

# Service call in yaml:
# service: python_script.storystore
# data:
#   useentity: ENT    # ENT is entity_id of entity to use for storing records, mandatory.
#   useattribute: ATT # ATT is attribute to use for storing records. Default is 'story'.
#   usestamp: STAMP   # STAMP sets the format to use for records. Default is full stamp, like 2023-03-26 13:47:55.4545.
# hms --> 13:47:55, hm --> 13:47
#   max: MAX          # MAX is the maximum of records to be stored. Default is 10
#   restore: STRING   # Attribute is set to STRING. To be used when HA is restarted.
#   logstring: STRING # STRING is the complete string to put after the timestamp in record.
#   logidstate: LIST  # LIST is an array of entity ids [ID1,ID2,ID3,...]. The string to put in record is formated as:
# ID1|STATE1|ID2|STATE2... Example record: 2023-03-26 13:47:55.4545|light.lamp|ON|sensor.temp|21.4;
#   logstate: LIST    # LIST is an array of entity ids [ID1,ID2,ID3,...]. The string to put in record is formated as:
# STATE1|STATE2|STATE3... Example record: 2023-03-26 13:47:55.4545|ON|21.4|4711;

# One and only one of restore, logstring, logidstate or logstate shall be provided.


# Detected problems: "data" is not defined  - but "data" is built in

useentity = data.get('useentity', '')
useattribute = data.get('useattribute', 'story')
usestamp = data.get('usestamp', 'full')
logidstate = data.get('logidstate', '')
logstate = data.get('logstate', '')
logstring = data.get('logstring', '')
restore = data.get('restore', '')
max = int(data.get('max', '10'))

# save all attributes but story
attrs = hass.states.get(useentity).attributes
result = {}
for key in attrs:
    if key != useattribute:
        result[key] = attrs[key]


stamp = f'{datetime.datetime.now()}'
if usestamp == 'hm':
    stamp = stamp[11:16]
elif usestamp == 'hms':
    stamp = stamp[11:19]
elif usestamp == 'raw':
    stamp = str(time.time()*1000)  # ms since 1970

if (restore != ''):  # recreate complete story attribute
    result[useattribute] = restore
else:
    if (logstring != ''):     # create new record from log
        story = f'{stamp}|{logstring}'
    elif (logidstate != ''):  # create new record with id|state from list
        story = stamp
        ix = 0
        while (ix < len(logidstate)):
            story = story + '|' + \
                logidstate[ix] + '|' + hass.states.get(logidstate[ix]).state
            ix = ix + 1
    elif (logstate != ''):  # create new record with state|state from list
        story = stamp
        ix = 0
        while (ix < len(logstate)):
            story = story + '|' + hass.states.get(logstate[ix]).state
            ix = ix + 1

    records = attrs.get(useattribute, '').split(';')
    n = 1  # count of records, 1 is the new one
    j = len(records)-1  # index of newest existing
    while (j >= 0) and (n < max):
        if (records[j] != ''):
            story = records[j] + ';' + story
        j = j-1
        n = n+1
    result[useattribute] = story

hass.states.set(useentity, stamp, result, True)

File: src/storystore.js
// skannea
// ============================================= Storystore
const storystore = {
  // ---------------------------------------------
  // 11:33:55|abra|kadabra;12:34:56|hocus|pocus;12:35:12|the end
  // [['11:33:55','abra','kadabra'],['11:34:15','hocus','pocus'],['11:38:57','the end']]
  // make one array per record
  textRows: function (data) {
    var inrecords = data.log.split(";");
    var outrecords = [];
    var i = 0; // starting with earliest record
    while (i < inrecords.length) {
      var ar = inrecords[i].split("|");
      outrecords.push(ar);
      i++;
    }
    return outrecords;
  },
  // ---------------------------------------------
  // 11:33:55|sensor.a|47.0;12:34:56|sensor.b|27.4;12:35:12|sensor.b|27.8
  // [{ stamp:'11:33:55',id:'sensor.a',state:'47.0' },{ stamp:'12:34:56',id:'sensor.b',state:'27.4' },{ stamp:'12:35:12',id:'sensor.b',state:'27.8' }]
  // assume stamp|id|state, make one object per record
  idstateRows: function (data) {
    var inrecords = data.log.split(";");
    var outrecords = [];
    var i = 0; // starting with earliest record
    while (i < inrecords.length) {
      var ar = inrecords[i].split("|");
      outrecords.push({ stamp: ar[0], id: ar[1], state: ar[2] });
      i++;
    }
    return outrecords;
  },
  // ---------------------------------------------
  // 11:33:55|sensor.a|47.0|sensor.b|27.5|sensor.c|0.5;12:34:56|sensor.a|47.1|sensor.b|27.4|sensor.c|0.5
  // { stamps:['11:33:55','12:34:56'], ids:[ 'sensor.a', 'sensor.b', 'sensor.c' ],  states:[['47.0','47.1'],['27.5','27.4'],['0.5','0.5']] }
  // assume stamp|id|state, sort records and make { stamps:[], ids:[], states:[] } where states[n][t]-->stamps[t], ids[n]
  idstateCols: function (data) {
    var records = data.log.split(";");
    var axes = { stamps: [], ids: [], states: [] };
    var i = 0; // starting with earliest record
    while (i < records.length) {
      var ar = records[i].split("|");
      axes.stamps.push(ar[0]);
      var j = 1; // index for parts of a record (ar)
      while (j < ar.length) {
        if (i == 0) axes.ids.push(ar[j]);
        j++;
        if (i == 0) axes.states[j / 2 - 1] = [];
        axes.states[j / 2 - 1].push(ar[j]);
        j++;
      }
      i++;
    }
    return axes;
  },
  // -----------------
  // 11:33:55|47.0|27.5|0.5;12:34:56|47.1|27.4|0.5
  // { stamps:['11:33:55','12:34:56'],  states:[['47.0','47.1'],['27.5','27.4'],['0.5','0.5']] }
  // assume stamp|state, sort records and make { stamps:[], states:[] } where states[n][t]-->stamps[t] (n-->id is known)
  stateCols: function (data) {
    var records = data.log.split(";");
    var axes = { stamps: [], states: [] };
    var i = 0; // starting with earliest record
    while (i < records.length) {
      var ar = records[i].split("|");
      axes.stamps.push(ar[0]);
      var j = 1; // index for parts of a record (ar)
      while (j < ar.length) {
        if (i == 0) axes.states[j - 1] = [];
        axes.states[j - 1].push(ar[j]);
        j++;
      }
      i++;
    }
    return axes;
  },
  // 2023-04-18 13:14:15.123456
  // 01234567890123456789012345
  // ---------------------------------------------
  to_hms: function (dates) {
    for (var i = 0; i < dates.length; i++)
      if (dates[i].length > 17)
        // assume long datetime
        dates[i] = dates[i].slice(11, 19);
    return dates;
  },
  // ---------------------------------------------
  to_hm: function (dates) {
    for (var i = 0; i < dates.length; i++)
      if (dates[i].length > 17) dates[i] = dates[i].slice(11, 16);
    return dates;
  },
  // ---------------------------------------------
  to_ymd: function (dates) {
    for (var i = 0; i < dates.length; i++)
      if (dates[i].length > 17) dates[i] = dates[i].slice(0, 10);
    return dates;
  },
  // ---------------------------------------------
  to_md: function (dates) {
    for (var i = 0; i < dates.length; i++)
      if (dates[i].length > 17) dates[i] = dates[i].slice(5, 10);
    return dates;
  },
};
